src/jvm/main/org/jetbrains/kotlinx/lincheck/TraceDebugger.kt

```diff
@@ -63,11 +90,19 @@ internal object TraceDebuggerInjections {
             .logLevel(LoggingLevel.OFF)
             .invocationTimeout(5 * 60 * 1000) // 5 mins
 
-        @Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
         val failure = lincheckOptions.checkImpl(instanceClass)
 
         val result = failure!!.results.threadsResults[0][0]
         if (result is ExceptionResult) throw result.throwable
+
+        if (!dumpTraceIntoFile.isNullOrEmpty() && failure.trace != null) {
+            val trace = constructTraceForPlugin(failure, failure.trace)
+            val dumpFile = File(dumpTraceIntoFile!!)
+            dumpFile.parentFile.mkdirs()
+            dumpFile.createNewFile()
+            dumpFile.writeText(trace.joinToString("\n"))
+            exitProcess(0)
```

Comment:
Why `exitProcess(0)` is required here? 
And why it is only called in one branch of the `if` ?

Reply:
This is strange. 

> the agent holds the thread for the debugger

Yes, the agent will execute the method under trace debugging in a loop to enable replays, but it should only happen if the special flag is set (which is set by the IDEA debugger). 

Moreover, this loop is run inside this line:
```
val failure = lincheckOptions.checkImpl(instanceClass)
```  

so if the code reached this point, it should have already exited the loop and should proceed normally.

Reply:
Could we maybe introduce a separate option for the agent --- the mode that we run the trace debugger in?

```
enum class TraceDebuggerMode { TRACE_COLLECTION, TRACE_DEBUGGING } 
``` 

And then
- use `TRACE_DEBUGGING` by default;
- pass `TRACE_COLLECTION` when running integration tests;
- call `exit(0)` only in `TRACE_COLLECTION` mode.

Reply:
Checked it again and everything works even without `exitProcess`. I wonder why and that's changed...